fe/fe-core/src/main/java/com/starrocks/qe/SimpleScheduler.java

```diff
@@ -108,130 +108,61 @@ public static TNetworkAddress getHost(long backendId,
+                                                 Reference<Long> backendIdRef) {
+        Backend node = getBackend(backendMap);
+        if (node != null) {
+            backendIdRef.setRef(node.getId());
+            return new TNetworkAddress(node.getHost(), node.getBePort());
         }
-        // no backend returned
         return null;
     }
 
-    public static ComputeNode getComputeNode(ImmutableMap<Long, ComputeNode> computeNodes) {
-        if (computeNodes == null) {
+    @Nullable
+    public static Backend getBackend(ImmutableMap<Long, Backend> nodeMap) {
+        if (nodeMap == null || nodeMap.isEmpty()) {
             return null;
         }
-        int computeNodeSize = computeNodes.size();
-        if (computeNodeSize == 0) {
+        return chooseNode(nodeMap.values().asList(), nextBackendHostId);
+    }
+
+    @Nullable
+    public static ComputeNode getComputeNode(ImmutableMap<Long, ComputeNode> nodeMap) {
+        if (nodeMap == null || nodeMap.isEmpty()) {
             return null;
         }
-        long id = nextComputeNodeId.getAndIncrement() % computeNodeSize;
+        return chooseNode(nodeMap.values().asList(), nextComputeNodeHostId);
+    }
 
-        List<Long> idToComputeNodeId = Lists.newArrayList();
-        idToComputeNodeId.addAll(computeNodes.keySet());
-        Long computeNodeId = idToComputeNodeId.get((int) id);
-        ComputeNode computeNode = computeNodes.get(computeNodeId);
+    @Nullable
+    private static <T extends ComputeNode> T chooseNode(ImmutableList<T> nodes, AtomicLong nextId) {
+        int index = (int) (nextId.getAndIncrement() % nodes.size());
+        T node = nodes.get(index);
+        if (node != null && node.isAlive() && !blacklistBackends.containsKey(node.getId())) {
+            return node;
+        }
 
-        if (computeNode != null && computeNode.isAlive() && !blacklistBackends.containsKey(computeNodeId)) {
-            return computeNode;
-        } else {
-            long candidateId = nextComputeNodeId.getAndIncrement() % computeNodeSize;  // get next candidate id
-            for (int i = 0; i < computeNodeSize; i++) {
-                LOG.debug("i={} candidatedId={}", i, candidateId);
-                if (candidateId == id) {
-                    continue;
-                }
-                Long candidatebackendId = idToComputeNodeId.get((int) candidateId);
-                LOG.debug("candidatebackendId={}", candidatebackendId);
-                ComputeNode candidateBackend = computeNodes.get(candidatebackendId);
-                if (candidateBackend != null && candidateBackend.isAlive()
-                        && !blacklistBackends.containsKey(candidatebackendId)) {
-                    return candidateBackend;
-                }
-                candidateId = nextComputeNodeId.getAndIncrement() % computeNodeSize;
-            }
+        // slow path
+        List<T> aliveNodes = nodes.stream()
```

Comment:
This means every time we will scan all the backends, for the big cluster with 100+ BEs, this may slow the query/load plan. How about first get the nextId from the Atomic nextId to a local variable, and then check the BE status according the local nextId, is not alive, increment the local nextId, and at last add the delta to the Atomic nextId.

Reply:
updated